File: Arrays & Strings/#169 - Majority Element - Easy/majorityElement.cpp
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int n = nums.size();
        int cnt = 0;
        int el;

        // Boyer-Moore Voting Algorithm
        for (int i = 0; i < n; i++) {
            if (cnt == 0) {
                cnt = 1;
                el = nums[i];
            } else if (nums[i] == el) {
                cnt++;
            } else {
                cnt--;
            }
        }

        // Optional: Verify if the candidate is actually the majority
        int cnt1 = 0;
        for (int i = 0; i < n; i++) {
            if (nums[i] == el)
                cnt1++;
        }
        if (cnt1 > n / 2)
            return el;
        return -1;
    }
};

File: Arrays & Strings/#169 - Majority Element - Easy/majorityElement.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        n = len(nums)
        cnt = 0
        el = None

        # Boyer-Moore Voting Algorithm
        for i in range(n):
            if cnt == 0:
                cnt = 1
                el = nums[i]
            elif nums[i] == el:
                cnt += 1
            else:
                cnt -= 1

        # Optional: Verify if el is actually the majority element.
        cnt1 = 0
        for i in range(n):
            if nums[i] == el:
                cnt1 += 1
        if cnt1 > n // 2:
            return el
        return -1